src/control/control.rs

```diff
@@ -0,0 +1,173 @@
+use crate::gamestate::*;
+use super::keyboard::*;
+use super::actions::Action;
+
+use std::sync::*;
+use std::time::Duration;
+use sdl2::*;
+
+const KEYHOLDDELAY: Duration = Duration::from_millis(30);
+
+#[derive(PartialEq, Clone, Copy)]
+enum Keystate {
+    None,
+    Press(Duration),
+    Tap,
+    Hold
+}
+
+pub struct Controller {
+    // what happens when you pour orange juice into milk?
+    k: Keyboard,
+    u: Keystate,
+    r: Keystate,
+    d: Keystate,
+    l: Keystate,
+}
+
+impl Controller {
+    pub fn new() -> Self {
+        Controller {
+            k: Keyboard::default(),
+            u: Keystate::None,
+            d: Keystate::None,
+            l: Keystate::None,
+            r: Keystate::None,
+        }
+    }
+    pub fn control(&mut self, pump: &Mutex<EventPump>, now: Duration, gamedata: Arc<Gamedata>, pid: usize) -> bool {
```

Comment:
This is kinda super ritardando to have now be a duration rather than an instant

Reply:
"Instants are always guaranteed to be no less than any previously measured instant when created, and are often useful for tasks such as measuring benchmarks or timing how long an operation takes.

"Note, however, that instants are not guaranteed to be steady. In other words, each tick of the underlying clock might not be the same length (e.g. some seconds may be longer than others). An instant may jump forwards or experience time dilation (slow down or speed up), but it will never go backwards.

"Instants are opaque types that can only be compared to one another. There is no method to get “the number of seconds” from an instant. Instead, it only allows measuring the duration between two instants (or comparing two instants)."

stfricku ritardando

Reply:
Ok are you actually an uberritardando... the `.elapsed` function that you use to get now (within gameloop) uses the same timing method as instant does, so your duration is no different, in terms of the steadiness of it
Read `man 2 clock_gettime`--rust either uses CLOCK_MONOTONIC or CLOCK_MONOTONIC_COARSE... both for instants and durations--because durations are gotten using instants
I said now is a ritardando name for an instant because of the .now() way to create an instant.
And I will double down on that by saying that you measuring the difference from the time since game start and now and subtracting from that the time since game start and current is just silly when you could instead just subtract the time of key press from the current time.

Reply:
the time since game start is a useful quantity to ensure that certain actions only occur at certain points in the game
now.elapsed can be used now but will need to be changed later.